Introduce checks to validate public API tests

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131390

This PR introduces a new sanity check for the public API tests in `.ci/pytorch/test.sh`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file OR an existing file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when a module underneath `torch/` cannot be imported.
* Runs this in CI as part just before the pre-existing FC / BC checks.

I've verified that re-introducing the bug that #131386 fixed causes the new check to fail:
![public_api_failure](https://github.com/user-attachments/assets/376ddef3-d14a-41f6-93e2-f935deb6555a)

jbschlosser committed:
Introduce linter to validate public API tests

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce linter to validate public API tests"


This PR introduces a new sanity check for the public API tests in `.github/scripts/check_public_API_tests.py`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when an existing file underneath `torch/` introduces an invalid public API function.
* Runs this on the lint shard in CI as part of the linter that validates test running.

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce linter to validate public API tests"


This PR introduces a new sanity check for the public API tests in `.github/scripts/check_public_API_tests.py`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when an existing file underneath `torch/` introduces an invalid public API function.
* Runs this on the lint shard in CI as part of the linter that validates test running.

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce linter to validate public API tests"


This PR introduces a new sanity check for the public API tests in `test/forward_backward_compatibility/check_public_API_tests.py`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when an existing file underneath `torch/` introduces an invalid public API function.
* Runs this in CI as part just before the pre-existing FC / BC checks.

[ghstack-poisoned]

jbschlosser commented on test/test_public_bindings.py:
```diff
@@ -594,7 +594,8 @@ def check_one_element(elem, modname, mod, *, is_public, is_all):
 
         for mod in pkgutil.walk_packages(torch.__path__, "torch."):
-            modname = mod.name
+            # TODO: Change this back after verifying the check catched the bug in CI
```
catched -> caught lmao

jbschlosser committed:
Update on "Introduce linter to validate public API tests"


This PR introduces a new sanity check for the public API tests in `test/forward_backward_compatibility/check_public_API_tests.py`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when an existing file underneath `torch/` introduces an invalid public API function.
* Runs this in CI as part just before the pre-existing FC / BC checks.
* Re-introduces a bug from the previous PR to validate that the new check catches the bug. **TODO: Revert this before merging!**

[ghstack-poisoned]

albanD commented on test/forward_backward_compatibility/check_public_API_tests.py:
```diff
@@ -0,0 +1,102 @@
+        "Success! 'test_correct_module_names' identified a new, invalid public API function."
+    )
+
+# Step 2. Make sure that the public API test "test_modules_can_be_imported" fails when an existing
+# file is modified to introduce an invalid public API function.
```
Huh, this test checks that "import your.module" doesn't crash :p 
What you describe here would be running test_correct_module_names again but after modifying an existing file right?

jbschlosser replied:
Hm I'm trying to follow the description for this task:
```
In particular we want to make sure that both these tests run and are not broken (this happened before during some CI infra update). These tests can be (and have been) silently broken/disabled before and we have no signal about it! A way to do this is to have the lint shard run the following: 1. create a bad file 2. make sure test_public_bindings.py -k test_correct_module_names fails 3. append bad API to existing file 4. make sure test_modules_can_be_imported (same file) fails 5. undo the changes.
```

In practice, I've seen that modifying `torch/nn/functional.py` with an invalid public function does indeed fail `test_modules_can_be_imported`, but do we care about testing this test and, if so, if this the right way to do it?

jbschlosser replied:
Updated to include the following test cases:
1. test_correct_module_names() fails for an invalid public API function in a new file
2. test_correct_module_names() fails for an invalid public API function in an existing file
3. test_modules_can_be_imported() fails when some module can't be imported

jbschlosser committed:
Update on "Introduce checks to validate public API tests"


This PR introduces a new sanity check for the public API tests in `test/forward_backward_compatibility/check_public_API_tests.py`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when an existing file underneath `torch/` introduces an invalid public API function.
* Runs this in CI as part just before the pre-existing FC / BC checks.

I've verified that re-introducing the bug that #131386 fixed causes the new check to fail:
![public_api_failure](https://github.com/user-attachments/assets/376ddef3-d14a-41f6-93e2-f935deb6555a)

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce checks to validate public API tests"


This PR introduces a new sanity check for the public API tests in `test/forward_backward_compatibility/check_public_API_tests.py`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when an existing file underneath `torch/` introduces an invalid public API function.
* Runs this in CI as part just before the pre-existing FC / BC checks.

I've verified that re-introducing the bug that #131386 fixed causes the new check to fail:
![public_api_failure](https://github.com/user-attachments/assets/376ddef3-d14a-41f6-93e2-f935deb6555a)

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce checks to validate public API tests"


This PR introduces a new sanity check for the public API tests in `test/forward_backward_compatibility/check_public_API_tests.py`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when an existing file underneath `torch/` introduces an invalid public API function.
* Runs this in CI as part just before the pre-existing FC / BC checks.

I've verified that re-introducing the bug that #131386 fixed causes the new check to fail:
![public_api_failure](https://github.com/user-attachments/assets/376ddef3-d14a-41f6-93e2-f935deb6555a)

[ghstack-poisoned]

albanD commented on test/forward_backward_compatibility/check_public_API_tests.py:
```diff
@@ -0,0 +1,108 @@
+    print("Creating a non-importable module...")
+    with open(os.path.join(d, "__init__.py"), "wb") as f:
+        f.write(b"invalid syntax garbage")
+        f.seek(0)
```
nit: not needed

albanD reviewed (approved):
SGTM!

jbschlosser committed:
Update on "Introduce checks to validate public API tests"


This PR introduces a new sanity check for the public API tests in `test/forward_backward_compatibility/check_public_API_tests.py`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when an existing file underneath `torch/` introduces an invalid public API function.
* Runs this in CI as part just before the pre-existing FC / BC checks.

I've verified that re-introducing the bug that #131386 fixed causes the new check to fail:
![public_api_failure](https://github.com/user-attachments/assets/376ddef3-d14a-41f6-93e2-f935deb6555a)

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce checks to validate public API tests"


This PR introduces a new sanity check for the public API tests in `test/forward_backward_compatibility/check_public_API_tests.py`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file OR an existing file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when a module underneath `torch/` cannot be imported.
* Runs this in CI as part just before the pre-existing FC / BC checks.

I've verified that re-introducing the bug that #131386 fixed causes the new check to fail:
![public_api_failure](https://github.com/user-attachments/assets/376ddef3-d14a-41f6-93e2-f935deb6555a)

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce checks to validate public API tests"


This PR introduces a new sanity check for the public API tests in `test/forward_backward_compatibility/check_public_API_tests.py`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file OR an existing file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when a module underneath `torch/` cannot be imported.
* Runs this in CI as part just before the pre-existing FC / BC checks.

I've verified that re-introducing the bug that #131386 fixed causes the new check to fail:
![public_api_failure](https://github.com/user-attachments/assets/376ddef3-d14a-41f6-93e2-f935deb6555a)

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce checks to validate public API tests"


This PR introduces a new sanity check for the public API tests in `test/forward_backward_compatibility/check_public_API_tests.py`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file OR an existing file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when a module underneath `torch/` cannot be imported.
* Runs this in CI as part just before the pre-existing FC / BC checks.

I've verified that re-introducing the bug that #131386 fixed causes the new check to fail:
![public_api_failure](https://github.com/user-attachments/assets/376ddef3-d14a-41f6-93e2-f935deb6555a)

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce checks to validate public API tests"


This PR introduces a new sanity check for the public API tests in `test/forward_backward_compatibility/check_public_API_tests.py`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file OR an existing file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when a module underneath `torch/` cannot be imported.
* Runs this in CI as part just before the pre-existing FC / BC checks.

I've verified that re-introducing the bug that #131386 fixed causes the new check to fail:
![public_api_failure](https://github.com/user-attachments/assets/376ddef3-d14a-41f6-93e2-f935deb6555a)

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce checks to validate public API tests"


This PR introduces a new sanity check for the public API tests in `test/forward_backward_compatibility/check_public_API_tests.py`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file OR an existing file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when a module underneath `torch/` cannot be imported.
* Runs this in CI as part just before the pre-existing FC / BC checks.

I've verified that re-introducing the bug that #131386 fixed causes the new check to fail:
![public_api_failure](https://github.com/user-attachments/assets/376ddef3-d14a-41f6-93e2-f935deb6555a)

[ghstack-poisoned]

albanD reviewed (approved):
SGTM !

jbschlosser committed:
Update on "Introduce checks to validate public API tests"


This PR introduces a new sanity check for the public API tests in `.ci/pytorch/test.sh`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file OR an existing file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when a module underneath `torch/` cannot be imported.
* Runs this in CI as part just before the pre-existing FC / BC checks.

I've verified that re-introducing the bug that #131386 fixed causes the new check to fail:
![public_api_failure](https://github.com/user-attachments/assets/376ddef3-d14a-41f6-93e2-f935deb6555a)

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
